### PR TITLE
[cxxmodule] Enable explicit modules

### PR DESCRIPTION
**Pull request by yamaguchi1024**
These changes includes adding `--cxxmodule` flag to genreflex, and
adding interface headers to classes.h

We are planning to automatically generate them
(https://github.com/cms-sw/cmssw/issues/25903), but for now we can
manually maintain them.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by smuzaffar** [comment]
@yamaguchi1024 , I would not suggest to test/merge this in cmssw master. No idea how root 6.12 , 6.14 and master IBs will behave with --cxxmodule flag.

**Comment by yamaguchi1024** [comment]
@smuzaffar , They will ignore the --cxxmodule flag, but I forgot to add <if CXXMODULE>, I will update this

**Comment by slava77** [comment]
-1

resolution(s) to the build errors and other update are expected, apparently

File: DataFormats/Common/src/classes.h [hunk 1]
```diff
@@ -1,34 +1,118 @@
...
 #include "DataFormats/Common/interface/TriggerResults.h"
-#include "DataFormats/Common/interface/Wrapper.h"
-#include "DataFormats/Common/interface/FillView.h"
-#include "DataFormats/Common/interface/DataFrame.h"
-#include "DataFormats/Common/interface/DataFrameContainer.h"
-#include "DataFormats/Common/interface/DetSetVectorNew.h"
-#include "DataFormats/Common/interface/ConstPtrCache.h"
-#include "DataFormats/Common/interface/BoolCache.h"
-#include "DataFormats/Common/interface/PtrVectorBase.h"
+#include "DataFormats/Common/interface/ValidHandle.h"
```
Comment (Dr15Jones) [inline DataFormats/Common/src/classes.h]: There is absolutely no reason to make a ROOT dictionary for ValidHandle. If `classes.h` is intended just for dictionary generation, putting all headers into the file doesn't seem like a good idea. If the intent is preparation for C++ modules, then, OK, I guess.

Comment (yamaguchi1024) [reply DataFormats/Common/src/classes.h]: I see. It is a preparation for C++ modules, but why ValidHandle.h is in interface/ if it is not supposed to be exposed to users? (Shouldn't it be under src if it was like an internal header?)

Comment (Dr15Jones) [reply DataFormats/Common/src/classes.h]: Up until now, we've only used classes.h to create the ROOT dictionaries we need for either I/O or for access to member functions of certain classes. `ValidHandle` is not necessary for either of those cases but it is needed for other code in CMSSW.

If classes.h is now being used as preparation for C++ modules then `ValidHandle` should be there.

**Comment by Dr15Jones** [comment]
@smuzaffar how should this be tested? Just running the regular tests?

**Comment by smuzaffar** [comment]
@Dr15Jones , I have changed this PR to be based on 10.6.CXXMODULE.X branch. This should be testing against CXXMODULE IB to make use of different build rules.

**Comment by smuzaffar** [comment]
please test

File: DataFormats/Math/interface/AVXVec.h [hunk 2]
```diff
@@ -1,6 +1,8 @@
 #ifndef DataFormat_Math_AVXVec_H
 #define DataFormat_Math_AVXVec_H
 
+#include "DataFormats/Math/interface/SSEVec.h"
```
Comment (smuzaffar) [inline DataFormats/Math/interface/AVXVec.h]: @yamaguchi1024 , is not it going to create a cyclic dependency (SSEVec.h includes AVXVec.h too)?

**Comment by smuzaffar** [comment]
@yamaguchi1024 , CXXMODULE IB build rules already add all missing packages/interface/*.h in to its src/classes.h. Do you still explicitly wants to add them?

File: DataFormats/Math/interface/AVXVec.h (hunk 2 above)
Comment (yamaguchi1024) [reply DataFormats/Math/interface/AVXVec.h]: It'll not be a cyclic dependency if headers were in the same module (SSWVec.h and AVXVec.h are both in DataFormatsMath_xr.pcm), but cyclic dependency occurs when headers in the different module are including each other. (Like https://github.com/cms-sw/cmssw/issues/25903#issuecomment-463298741 case)

**Comment by yamaguchi1024** [comment]
This error is because Clang is finding two modulemaps defining the same module in its INCLUDE_PATH (`/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_CXXMODULE_X_2019-02-21-2300/src/module.modulemap` and `./module.modulemap`). I feel IB's paths are not intended to be added to INCLUDE_PATH at build time (If someone accidentally delete one header in the current build but the compiler still finds the old header from IB paths?), but what do you think?

**Comment by yamaguchi1024** [comment]
> @yamaguchi1024 , CXXMODULE IB build rules already add all missing packages/interface/*.h in to its src/classes.h. Do you still explicitly wants to add them?

Oh great! Then I guess we don't need classes.h modifications.

**Comment by slava77** [comment]
@yamaguchi1024 
please clarify on the status/plans for this PR.
Thank you.

**Comment by yamaguchi1024** [comment]
@slava77 Thanks for the heads-up!

Closing PR as it is outdated, we have now auto-generated module.modulemap and classes.h thanks to @smuzaffar . From now when we want to enable C++ Modules for CMS libraries, all we need to do is to add
```
<release name="_CXXMODULE_"><flags   GENREFLEX_ARGS="--deep --cxxmodule"/></release>
```
to BuildFile.xml.

**Pull request closed without merging**